diff --git a/tallow.c b/tallow.c
index 784373b..502a48c 100644
--- a/tallow.c
+++ b/tallow.c
@@ -89,10 +89,15 @@ static void block(struct tallow_struct *s)
 		free(s->ip);
 		free(s);
 	} else {
-		struct tallow_struct *p = s;
-		s = s->next;
-		free(p->ip);
-		free(p);
+		struct tallow_struct *p = head;
+
+		while (p) {
+			if (p->next == s) 
+				p->next = s->next;
+			p = p->next;
+		}
+		free(s->ip);
+		free(s);
 	}
 }
 
@@ -125,6 +130,9 @@ static void find(char *ip)
 	struct tallow_struct *n;
 	struct tallow_struct *w = whitelist;
 
+	if (!ip)
+		return;
+
 	/*
 	 * not validating the IP address format here, just
 	 * making sure we're not passing special characters
@@ -353,6 +361,14 @@ int main(void)
 				find(t);
 			}
 
+
+			if (strstr(m, "MESSAGE=Received disconnect from ")) {
+				t = strtok(m, " ");
+				for (i = 0; i < 4; i++)
+					t = strtok(NULL, " ");
+				find(t);
+			}
+
 			free(m);
 		}
 	}
